pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go

```diff
@@ -0,0 +1,391 @@
+	resourceName := chi.URLParam(r, "resourceName")
+	resourceType := chi.URLParam(r, "resourceType")
+	resourceGroupName := chi.URLParam(r, "resourceGroupName")
+
+	resourceID := strings.TrimPrefix(r.URL.Path, "/admin")
+
+	doc, err := f.dbOpenShiftClusters.Get(ctx, resourceID)
+	switch {
+	case cosmosdb.IsErrorStatusCode(err, http.StatusNotFound):
+		return api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", resourceType, resourceName, resourceGroupName)
+	case err != nil:
+		return err
+	}
+
+	k, err := f.kubeActionsFactory(log, f.env, doc.OpenShiftCluster)
+	if err != nil {
+		return err
+	}
+	e := &etcdrenew{
+		log:           log,
+		k:             k,
+		secretNames:   nil,
+		mode:          "renew",
+		backupSecrets: make(map[string][]byte),
+		lastRevision:  0,
+	}
+
+	if err = e.validateClusterVersion(ctx); err != nil {
+		return err
+	}
+
+	// Fetch secretNames using nodeNames
+	masterNodeNames, err := fetchNodeNames(ctx, k, log)
+	if err != nil {
+		return err
+	}
+	if len(masterNodeNames) != 3 {
+		return api.NewCloudError(http.StatusForbidden, api.CloudErrorCodeForbidden, "", "The cluster doesn't have 3 master nodes")
+	}
+
+	for _, nodeName := range masterNodeNames {
+		for _, prefix := range []string{"etcd-peer-", "etcd-serving-", "etcd-serving-metrics-"} {
+			e.secretNames = append(e.secretNames, prefix+nodeName)
+		}
+	}
+
+	if err = e.validate(ctx); err != nil {
+		return err
+	}
+	// backup and delete etcd secrets
+	if err = e.backupAndDelete(ctx); err != nil {
+		return err
+	}
+
+	if err = e.isRenewed(ctx); err != nil {
+		e.mode = "recovery"
+	} else {
+		e.mode = "renewed"
+	}
+
+	if e.mode == "renewed" {
+		if err = e.validate(ctx); err != nil {
+			return err
+		}
+		e.log.Infoln("Done")
+	}
+
+	if e.mode == "recovery" {
+		e.log.Println("Attempting to recover from backup")
+		if err = e.recoverEtcdSecrets(ctx); err != nil {
+			return err
+		}
+		e.log.Infoln("Recovered")
+	}
+
+	return nil
+}
+
+func (e *etcdrenew) validateClusterVersion(ctx context.Context) error {
+	e.log.Infoln("validating cluster version now")
+	rawCV, err := e.k.KubeGet(ctx, "clusterversion", "", "version")
+	if err != nil {
+		return err
+	}
+	cv := &configv1.ClusterVersion{}
+	err = codec.NewDecoderBytes(rawCV, &codec.JsonHandle{}).Decode(cv)
+	if err != nil {
+		return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", fmt.Sprintf("failed to decode clusterversion, %s", err.Error()))
+	}
+	clusterVersion, err := version.GetClusterVersion(cv)
+	if err != nil {
+		return err
+	}
+	// ETCD ceritificates are autorotated by the operator when close to expiry for cluster running 4.9+
+	if !clusterVersion.Lt(version.NewVersion(4, 9)) {
+		return api.NewCloudError(http.StatusForbidden, api.CloudErrorCodeForbidden, "", "etcd certificate renewal is not needed for cluster running version 4.9+")
```

Comment:
Hey Srinivas. 4.9 is currently already considered EOL, correct? Can you elaborate a bit on how this action would work as our first suggestion for 4.9 clusters with a support issue should be to upgrade. Would this GA be used prior in order to facilitate said upgrade?

Reply:
Hey Jory! TL's made the decision that we should do this pro-actively even though we know these clusters are technically out of support. The reasoning was that recovering the clusters after they expired took 2 people 6 hours, and pro-actively rotating the certificates by hand took 2 people 3 minutes. We reasoned that creating an automation that allows 1 person to accomplish this (no need to elevate permissions) was the absolute best route, and so we created an initiative and several epics for this. See: https://issues.redhat.com/browse/ARO-3983